tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedTestOperations/UnifiedBulkWriteOperation.cs

```diff
@@ -0,0 +1,274 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+
+namespace MongoDB.Driver.Tests.Specifications.unified_test_format.UnifiedTestOperations
+{
+    public class UnifiedBulkWriteOperation : IUnifiedTestOperation
+    {
+        private IMongoCollection<BsonDocument> _collection;
+        private List<WriteModel<BsonDocument>> _requests;
+        private BulkWriteOptions _options;
+        private IClientSessionHandle _session;
+
+        public UnifiedBulkWriteOperation(
+            IMongoCollection<BsonDocument> collection,
+            List<WriteModel<BsonDocument>> requests,
+            BulkWriteOptions options,
+            IClientSessionHandle session)
+        {
+            _collection = collection;
+            _requests = requests;
+            _options = options;
+            _session = session;
+        }
+
+        public OperationResult Execute(CancellationToken cancellationToken)
+        {
+            BulkWriteResult<BsonDocument> result;
+
+            try
+            {
+                if (_session == null)
+                {
+                    result = _collection.BulkWrite(_requests, _options);
+                }
+                else
+                {
+                    result = _collection.BulkWrite(_session, _requests, _options);
+                }
+            }
+            catch (Exception ex)
+            {
+                return new OperationResult(ex);
+            }
+
+            return new UnifiedBulkWriteOperationResultConverter().Convert(result);
+        }
+
+        public async Task<OperationResult> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            BulkWriteResult<BsonDocument> result;
+
+            try
+            {
+                if (_session == null)
+                {
+                    result = await _collection.BulkWriteAsync(_requests, _options);
+                }
+                else
+                {
+                    result = await _collection.BulkWriteAsync(_session, _requests, _options);
+                }
+            }
+            catch (Exception ex)
+            {
+                return new OperationResult(ex);
+            }
+
+            return new UnifiedBulkWriteOperationResultConverter().Convert(result);
+        }
+    }
+
+    public class UnifiedBulkWriteOperationBuilder
```

Comment:
This class clearly shows why the separation into operation and builder is beneficial. There's a significant amount of code not involved in operation execution which otherwise would reside in the operation class.

Reply:
Nice. I like it.